Improve the type clash error message

Julow opened the pull request:
Related: https://github.com/ocaml/ocaml/issues/12152

The second commit re-use the mechanism implemented in https://github.com/ocaml/ocaml/pull/11679 for type clash errors:

    Error: The expression '43' has type int
           but an expression was expected of type int32

The third commit change the denomination for some expressions instead of the generic "expression".
For example:

    The constant '42' has type ...

    This field has type ...

This is similar to the previous PR (https://github.com/ocaml/ocaml/pull/11679):

    The function 'foo' has type ...

This could be extended to most expressions (eg. `object`, `unnamed function`, `tuple`, `match expression`, etc..).

Julow committed:
Implement 'exp_is_nominal' on Parsetree

This function can be used to improve the type clash error message, which
uses the parsetree instead of the typedtree to avoid loosing some
information from the source code (eg. how constants are printed).

This function is used for the "this function ..." message, which can be
easily changed to use untyped expression.

Julow committed:
Print short expressions in type error messages

Re-use the mechanism implemented for the "This function" message in
type clash errors to print the original expression if possible.

Type error messages now look like this:

    Error: The expression '43' has type int
           but an expression was expected of type int32

The argument 'unify_exp ~sexp' is made non-optional as this isn't too
intrusive and ensures the new error is implemented in more cases.

Julow commented:
This is request for comment. I don't mind dropping the third commit or giving up the entire idea.

gasche commented:
Interesting, thanks!

I think that I disagree with some of the category names. For example "The method `x#m`" reads wrong to me, `x#m` is a method call, `m` is the method. I suspect that I don't like "record field" for the same reason, but I wasn't able to find an example in the testsuite to make an opinion. Could you ensure that all special cases are covered by at least one test?

Julow committed:
Update Changes

Julow commented:
A more correct denomination might be "method call", "record field access". I don't think length is a problem at this point.

Perhaps the message could print "The method 'm' has type" and remove the `x#` part, that could not be nominal in some cases ?

Julow commented on testsuite/tests/typing-misc/exp_denom.ml:
```diff
@@ -0,0 +1,72 @@
+Line 3, characters 18-22:
+3 | let _ = print_int Cons
+                      ^^^^
+Error: This expression has type v but an expression was expected of type int
```
There's a bug here.

Julow commented on Changes:
```diff
@@ -17,6 +17,13 @@ Working version
 
 ### Compiler user-interface and warnings:
 
+- #12182: Improve the type clash error message.
+  For example, this message:
+    This expression has type ...
+  is changed into:
+    The constant '42' has type ...
+  (Jules Aguillon, review by Gabriel Scherer)
```
The change entry is perhaps too long ?

gasche replied:
Fine with me. I like descriptive change entries.

Octachron commented:
I believe that using `the method 'm'` or `the field 'f'` doesn't work because both don't have a type on their own. Consider for instance:
```ocaml
type ('a,'b,'c) t = { x:'a; y:'b; z:'c }
let f r = let a = r.z (1 + r.x) ("y" ^ r.y) in [None; r.z]
```
It would be confusing to state that
```ocaml
Error: The field 'z' has type int -> string -> 'a
       but an expression was expected of type 'b option
```
when there is no such constraints on the field `z`  in general.

gasche commented:
Do we want to revisit this or close it?

Octachron commented:
Overall, I like the change; and if @Julow has the time I would be happy to revisit it.

Julow commented:
Hi, I would be happy to revisit it. I also don't mind if this is closed even after I put some time into it.

Julow commented:
> I believe that using the method 'm' or the field 'f' doesn't work because both don't have a type on their own. Consider for instance:

To anchor the error message in the context, we could say "the method 'x#m'" or "the field 'r.f'" ?

Octachron commented:
Possibly, but then we have the issue that `the field r.f` reads a bit strange when the field is `f`.
Another option would be to restructure the error message from
```
"%a has type" "The field f"
```
to
```
"In the record projection* `r.f`, the field `f` has type" (* projection => access? *)
```
(which also has the advantage of being localisable)

Octachron commented on typing/typecore.ml:
```diff
@@ -5689,6 +5688,38 @@ let type_clash_of_trace trace =
+let report_this_exp denom ppf exp =
+  let denom = match denom with Some d -> d | None -> exp_denom exp in
+  if Pprintast.exp_is_nominal exp then
+    fprintf ppf "The %s '%a'" denom Pprintast.expression exp
```
The expression can now be quoted with `Style.as_inline_code`.

Julow replied:
I changed to the new quoting mechanism and improved quoting for string constants. I also made longer strings not nominal to avoid ridiculously long error messages.

Julow committed:
Denomination for multi-keyword expressions

Add denominations for expressions like `for`, `while`, `if`, `match`,
etc.. that are often spanning several lines.

It acts as a confirmation that the error is about the containing
expression rather than a nested one as the printed code is sometimes
harder to read.

There's no test for some of them (`try`, `if`) but they'll become
easily reachable if the denomination is used in other messages.

Julow commented:
With the last commits, a denomination is used for multi-keyword expressions like `for`, `while`, `if`, `match`.
It's a useful confirmation that the error is about the containing expression rather than a nested one, for which the printed code is hard to read when spanning several lines. There's no example for some of these but this would become useful if other messages use the new mechanism, notably `Apply_non_function` and `Not_a_function`.

What do you think of applying this change to the other messages starting in "This expression" ?

I worry that adding an intro to the message makes reading harder as it's of irregular length. What do you think of these:
Simpler and still localisable: (currently implemented in this branch)

    The field access "r.f" has type ...

More correct:

    The field "f" has type ...
    in the context of the field access "r.f"

Perhaps more uses of the context could be found ?

Octachron commented on testsuite/tests/typing-misc/exp_denom.ml:
```diff
@@ -79,3 +79,24 @@ Line 1, characters 18-23:
+Line 1, characters 14-36:
+1 | let _ : int = while false do () done
+                  ^^^^^^^^^^^^^^^^^^^^^^
+Error: This while expression has type "unit"
```
We want `inline code` quote here is since `while` is not part of the grammatical structure of the sentence.

Julow replied:
Implemented!

Octachron commented:
Using `field access` and `method call` is fine with me.  I think it is better to keep this PR simple, we can think later if we want to add more context. 

P.S: Note that those messages are not localizable, the issue is that composing error messages by fragments like `%a has type` makes the hypothesis that the subject and the rest of the sentence can be translated independently. This may be not be the case if for instance the verb conjugation depends on the subject gender.

Octachron commented on typing/typecore.ml:
```diff
@@ -415,12 +415,12 @@ let unify_pat_types_return_equated_pairs ~refine loc penv ty ty' =
+(** [sexp_for_hint] is used by error messages to report literals in their
     original formatting *)
-let unify_pat ?sdesc_for_hint env pat expected_ty =
+let unify_pat ?sexp_for_hint env pat expected_ty =
```
This PR is still only using the description for hints in the pattern case. I propose to keep the code unchanged.

Julow replied:
Done.

Octachron commented on typing/typecore.ml:
```diff
@@ -3114,14 +3114,14 @@ let name_cases default lst =
+(** [sexp_for_hint] is used by error messages to report literals in their
     original formatting *)
-let unify_exp ?sdesc_for_hint env exp expected_ty =
+let unify_exp ?sexp_for_hint env exp expected_ty =
```
The  `sexp_for_hint` is used for more than hints now. I think it would be better to rename it to `sexp` and make it non-optional. That will fix the missing constructors case.

Julow replied:
Thanks, this fixed many cases indeed.

Octachron commented on typing/typecore.ml:
```diff
@@ -6345,6 +6344,63 @@ let type_clash_of_trace trace =
+    - "This <denom> ..."
+    - "The <denom> 'foo' ..." *)
+let pp_exp_denom ppf exp =
+  let f = fprintf ppf "%s" in
```
`pp_print_string`?

Octachron commented on typing/typecore.ml:
```diff
@@ -6345,6 +6344,63 @@ let type_clash_of_trace trace =
+    | Pexp_constant { pconst_desc = Pconst_string (s, _, None); _ } ->
+        (* Avoid adding quotes around a quoted string constant but make sure to
+           use consistent inline code style. *)
+        Format.pp_print_string ppf (String.escaped s)
```
This is only an issue with dumb terminal and in the testsuite, isn't it? Moreover using `String.escaped` does not mesh well with non-ascii contents. I would propose to skip this step.

Julow replied:
You are right, making the strings nominal is problematic. I removed it.

Octachron commented on typing/typecore.ml:
```diff
@@ -6395,13 +6451,15 @@ let report_partial_application = function
 
 let report_expr_type_clash_hints exp diff =
   match exp with
-  | Some (Pexp_constant const) -> report_literal_type_constraint const diff
```
`match exp.pexp_desc with` ?

Julow replied:
Improved.

Octachron commented on typing/typecore.ml:
```diff
@@ -6345,6 +6344,63 @@ let type_clash_of_trace trace =
+        let d = match denom with Some d -> d | None -> "expression" in
+        fprintf ppf "%s" d
+      in
+      report_general_this_exp denom ppf
```
It seems to me that the two `report_this_exp` function could be merged without loss of readability.

Julow replied:
Do you mean change `report_this_pexp` to take an option `pexp` ?

Octachron replied:
Yes, I was thinking of making the choice of denomination in one place:
```ocaml
let report_this_pexp_opt denom ppf exp =
  let denom ppf =
    match denom, exp with
    | Some d, _ -> fprintf ppf "%s" d
    | None, Some exp -> pp_pexp_denom ppf exp
    | None, None -> fprintf ppf "expression"
  in
  let nexp = Option.bind exp Pprintast.Doc.(nominal_exp_doc longident) in
  match nexp with
  | Some nexp -> fprintf ppf "The %t %a" denom pp_doc nexp
  | _ -> fprintf ppf "This %t" denom
  ```

Octachron replied:
Looking at the call sites, we could also make the sentence more readable by adding the `has type` suffix.

Julow replied:
You removed the wrapping of `nexp` with `Style.as_inline_code` in your snippet. Is that intended ?

Octachron replied:
No, that was a mistake.

Julow replied:
Thanks, I've changed the code.

Octachron commented:
There is non-trivial conflict with #13169 (because I didn't convert most of the AST printers which are generally used to print immediately). Assuming that #13169 is merged first, @Julow don't hesitate to ping for fixing the conflict.

Julow commented on testsuite/tests/tool-toplevel/multi_phrase_line.compilers.reference:
```diff
@@ -21,8 +21,8 @@ Error: Syntax error
                     ^^^^
-Error: This expression has type "bool" but an expression was expected of type
-         "int"
+Error: The constructor "\#true" has type "bool"
```
This could be improved.

Octachron replied:
This looks like an instance of #13263.

Julow replied:
I can have a look at this issue later.

Julow reviewed (commented):
Rebased and ready for an other review. Thanks for waiting!

Octachron commented on parsing/pprintast.ml:
```diff
@@ -136,6 +136,48 @@ module Doc = struct
+     - Do not contain spaces when printed.
+     - Is a constant that is short enough.
+  *)
+  let nominal_exp_doc lid t =
```
I have the impression that we don't need to delay the dependency on `longident` anymore?
Also since the function is in the `Doc` submodule, the `_doc` suffix seems redundant.

Julow replied:
Done!

Octachron reviewed (approved):
The PR looks good to me: directly quoting the conflicting part of the source code should help readers to locate the source of the type error.

This should be a noticeable  improvement, which only requires a small and localized change in `Typecore`.

Octachron commented:
@Julow , do you wish to clean the history in order to preserve the individual message commits? Otherwise, I would squash the PR and merge the message commits.

Julow commented:
Thanks :) Yes, I can clean up the history.

Julow committed:
More precise denomination for some expressions

Clarify type clash error messages by printing the kind of the reported
expression.

Currently implemented for expression that are likely 'nominal' but not
necessarily.

Example:

    The constant "42" has type ...

    This field has type ...

This is similar to the `Apply_non_function` error:

    The function "foo" has type ...

    This function has type ...
